src/libraries/System.Security.Cryptography.Encoding/tests/Cbor.Tests/CborReaderTests.Array.cs

```diff
@@ -0,0 +1,272 @@
+            Assert.Equal(expectedLength, (int)length!.Value);
+
+            for (int i = 0; i < actualLength; i++)
+            {
+                ulong? innerLength = reader.BeginReadArray();
+                Assert.Equal(1, (int)innerLength!.Value);
+                reader.ReadInt64();
+                reader.EndReadArray();
+            }
+
+            Assert.Throws<FormatException>(() => reader.ReadInt64());
+        }
+
+        [Fact]
+        public static void BeginReadArray_EmptyBuffer_ShouldThrowFormatException()
+        {
+            byte[] encoding = Array.Empty<byte>();
+            var reader = new CborReader(encoding);
+
+            Assert.Throws<FormatException>(() => reader.BeginReadArray());
+        }
+
+        [Theory]
+        [InlineData("00")] // 0
+        [InlineData("20")] // -1
+        [InlineData("40")] // empty byte string
+        [InlineData("60")] // empty text string
+        [InlineData("f6")] // null
+        [InlineData("a0")] // {}
+        [InlineData("f97e00")] // NaN
+        [InlineData("fb3ff199999999999a")] // 1.1
+        public static void BeginReadArray_InvalidType_ShouldThrowInvalidOperationException(string hexEncoding)
+        {
+            byte[] data = hexEncoding.HexToByteArray();
+            var reader = new CborReader(data);
+            Assert.Throws<InvalidOperationException>(() => reader.BeginReadArray());
+        }
+
+        [Theory]
+        // Invalid initial bytes with array major type
+        [InlineData("9c")]
+        [InlineData("9d")]
+        [InlineData("9e")]
+        // valid initial bytes missing required definite length data
+        [InlineData("98")]
+        [InlineData("9912")]
+        [InlineData("9a000000")]
+        [InlineData("9b00000000000000")]
+        public static void BeginReadArray_InvalidData_ShouldThrowFormatException(string hexEncoding)
+        {
+            byte[] data = hexEncoding.HexToByteArray();
+            var reader = new CborReader(data);
+
+            Assert.Throws<FormatException>(() => reader.BeginReadArray());
+        }
+    }
+
+    static class ArrayReaderHelper
+    {
+        public static void ReadArray(this CborReader reader, params object[] expectedValues)
+        {
+            Assert.Equal(CborReaderState.StartArray, reader.Peek());
+
+            ulong? length = reader.BeginReadArray();
+
+            Assert.NotNull(length);
+            Assert.Equal(expectedValues.Length, (int)length!.Value);
+
+            foreach (object value in expectedValues)
+            {
+                switch (value)
+                {
+                    case int expected:
+                        if (expected >= 0)
+                        {
+                            Assert.Equal(CborReaderState.UnsignedInteger, reader.Peek());
+                        }
+                        else
+                        {
+                            Assert.Equal(CborReaderState.NegativeInteger, reader.Peek());
```

Comment:
If something goes wrong this is just going to say `Expected: NegativeInteger, Actual: TextString`.  If you think that's actionable enough without the state being recorded with it OK. (It probably is, I'm just raising the concern.)

Reply:
I presume you mean passing a user message to the assertion that prints the current read context? I can see how the current setup produce confusing error messages in CI, but given that xunit assertions haven't been designed for this kind of use, perhaps it might make more sense to replace these particular assertions with throwing exceptions?

Reply:
> perhaps it might make more sense to replace these particular assertions with throwing exceptions?

That could work.  I'm just imagining a situation where later ReadBoolean with a false value ends up not incrementing the position and want the tests to help guide us to the problem.  If you feel (and it's entirely a subjective call) that it's something easily determined, then it can be fine as-is.

It's less of a problem now that we can usually hit F5 and debug them; but I've definitely had sensible-looking helpers be frustrating and make me add a lot of Console.WriteLines to debug failures on systems that I only have SSH access to.

Reply:
I wouldn't expect the CBOR implementation to produce platform sensitive unit testing scenaria. So I'd probably just assume this is good enough until proven wrong.